File: CodeQualityCheckerPython/фИНАЛ ПП/CodeQualityChecker/CodeQualityChecker/Services/GitLabService.cs
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;
using Newtonsoft.Json.Linq;

public class GitLabService
{
    private readonly HttpClient _httpClient;
    private readonly IConfiguration _config;

    public GitLabService(HttpClient httpClient, IConfiguration config)
    {
        _httpClient = httpClient;
        _config = config;
    }

    /// <summary>
    /// Извлекает информацию о репозитории из GitLab URL
    /// </summary>
    public (string owner, string repo, string branch, string path) ParseGitLabUrl(string url)
    {
        url = url.TrimEnd('/');

        // Форматы URL:
        // https://gitlab.com/owner/repo
        // https://gitlab.com/owner/repo/-/tree/branch
        // https://gitlab.com/owner/repo/-/blob/branch/path/to/file

        var uri = new Uri(url);
        var pathSegments = uri.AbsolutePath.Trim('/').Split('/');

        if (pathSegments.Length < 2)
            throw new ArgumentException("Неверный формат URL GitLab репозитория");

        // Ищем позицию repo
        var repoIndex = 0;
        for (int i = 0; i < pathSegments.Length; i++)
        {
            if (pathSegments[i] != "-" && pathSegments[i] != "")
            {
                repoIndex = i;
                break;
            }
        }

        var owner = pathSegments[repoIndex];
        var repo = pathSegments[repoIndex + 1].Replace(".git", "");
        var branch = "main";
        var path = "";

        // Ищем tree или blob
        var treeIndex = -1;
        for (int i = 0; i < pathSegments.Length; i++)
        {
            if (pathSegments[i] == "tree")
            {
                treeIndex = i;
                if (i + 1 < pathSegments.Length)
                    branch = pathSegments[i + 1];
                break;
            }
            else if (pathSegments[i] == "blob")
            {
                treeIndex = i;
                if (i + 1 < pathSegments.Length)
                    branch = pathSegments[i + 1];
            }
        }

        if (treeIndex >= 0)
        {
            var pathStart = treeIndex + 2; // После tree/branch или blob/branch
            if (pathStart < pathSegments.Length)
            {
                path = string.Join("/", pathSegments.Skip(pathStart));
            }
        }

        return (owner, repo, branch, path);
    }

    /// <summary>
    /// Кодирует путь для GitLab API
    /// </summary>
    private string EncodeProjectPath(string owner, string repo)
    {
        return Uri.EscapeDataString($"{owner}/{repo}");
    }

    /// <summary>
    /// Получает содержимое файла из репозитория
    /// </summary>
    public async Task<string> GetFileContentAsync(string owner, string repo, string path, string branch, string? token = null)
    {
        var projectPath = EncodeProjectPath(owner, repo);
        var encodedPath = Uri.EscapeDataString(path);
        
        var url = $"https://gitlab.com/api/v4/projects/{projectPath}/repository/files/{encodedPath}?ref={branch}";
        
        var request = new HttpRequestMessage(HttpMethod.Get, url);
        
        if (!string.IsNullOrEmpty(token))
            request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", token);

        var response = await _httpClient.SendAsync(request);
        response.EnsureSuccessStatusCode();

        var json = JObject.Parse(await response.Content.ReadAsStringAsync());
        var content = json["content"]?.ToString();
        
        if (string.IsNullOrEmpty(content))
            throw new Exception("Не удалось получить содержимое файла");

        // Декодируем base64
        var decoded = Encoding.UTF8.GetString(Convert.FromBase64String(content));
        return decoded;
    }

    /// <summary>
    /// Получает список файлов в директории
    /// </summary>
    public async Task<List<GitFileInfo>> GetDirectoryContentsAsync(string owner, string repo, string path, string branch, string? token = null)
    {
        var projectPath = EncodeProjectPath(owner, repo);
        
        var url = string.IsNullOrEmpty(path)
            ? $"https://gitlab.com/api/v4/projects/{projectPath}/repository/tree?ref={branch}"
            : $"https://gitlab.com/api/v4/projects/{projectPath}/repository/tree?path={Uri.EscapeDataString(path)}&ref={branch}";

        var request = new HttpRequestMessage(HttpMethod.Get, url);
        
        if (!string.IsNullOrEmpty(token))
            request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", token);

        var response = await _httpClient.SendAsync(request);
        response.EnsureSuccessStatusCode();

        var json = JArray.Parse(await response.Content.ReadAsStringAsync());
        var files = new List<GitFileInfo>();

        foreach (var item in json)
        {
            files.Add(new GitFileInfo
            {
                Name = item["name"]?.ToString() ?? "",
                Path = item["path"]?.ToString() ?? "",
                Type = item["type"]?.ToString() ?? "blob",
                Size = 0
            });
        }

        return files;
    }

    /// <summary>
    /// Создает коммит с исправлениями и Merge Request
    /// </summary>
    public async Task<GitCommitResult> CreateCommitWithFixesAsync(
        string owner, string repo, string targetBranch,
        string filePath, string originalContent, string fixedContent,
        string commitMessage, string mrDescription, string? token = null)
    {
        if (string.IsNullOrEmpty(token))
            throw new ArgumentException("Требуется токен для создания коммита");

        var projectPath = EncodeProjectPath(owner, repo);
        var featureBranch = $"code-quality-fix-{DateTime.Now:yyyyMMddHHmmss}";

        // Создаем новую ветку
        await CreateBranchAsync(projectPath, featureBranch, targetBranch, token);

        // Создаем коммит с исправлением
        await CreateCommitAsync(projectPath, featureBranch, filePath, fixedContent, commitMessage, token);

        // Создаем Merge Request
        var mrIid = await CreateMergeRequestAsync(projectPath, featureBranch, targetBranch, commitMessage, mrDescription, token);

        return new GitCommitResult
        {
            Success = true,
            BranchName = featureBranch,
            PullRequestUrl = $"https://gitlab.com/{owner}/{repo}/-/merge_requests/{mrIid}",
            PullRequestNumber = mrIid
        };
    }

    /// <summary>
    /// Оставляет комментарий к Merge Request
    /// </summary>
    public async Task<bool> AddMergeRequestCommentAsync(
        string owner, string repo, int mrIid,
        string comment, string? token = null)
    {
        if (string.IsNullOrEmpty(token))
            throw new ArgumentException("Требуется токен для добавления комментария");

        var projectPath = EncodeProjectPath(owner, repo);
        
        var request = new HttpRequestMessage(HttpMethod.Post,
            $"https://gitlab.com/api/v4/projects/{projectPath}/merge_requests/{mrIid}/notes");
        
        request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", token);
        request.Content = new StringContent(
            JsonSerializer.Serialize(new { body = comment }),
            Encoding.UTF8,
            "application/json");

        var response = await _httpClient.SendAsync(request);
        return response.IsSuccessStatusCode;
    }

    private async Task CreateBranchAsync(string projectPath, string branchName, string baseBranch, string token)
    {
        var request = new HttpRequestMessage(HttpMethod.Post,
            $"https://gitlab.com/api/v4/projects/{projectPath}/repository/branches");
        
        request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", token);
        request.Content = new StringContent(
            JsonSerializer.Serialize(new
            {
                branch = branchName,
                @ref = baseBranch
            }),
            Encoding.UTF8,
            "application/json");

        var response = await _httpClient.SendAsync(request);
        response.EnsureSuccessStatusCode();
    }

    private async Task CreateCommitAsync(string projectPath, string branch, string path,
        string content, string message, string token)
    {
        var request = new HttpRequestMessage(HttpMethod.Post,
            $"https://gitlab.com/api/v4/projects/{projectPath}/repository/commits");
        
        request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", token);
        
        var actions = new[]
        {
            new
            {
                action = "update",
                file_path = path,
                content = Convert.ToBase64String(Encoding.UTF8.GetBytes(content))
            }
        };

        var body = new
        {
            branch = branch,
            commit_message = message,
            actions = actions
        };

        request.Content = new StringContent(
            JsonSerializer.Serialize(body),
            Encoding.UTF8,
            "application/json");

        var response = await _httpClient.SendAsync(request);
        response.EnsureSuccessStatusCode();
    }

    private async Task<int> CreateMergeRequestAsync(string projectPath,
        string sourceBranch, string targetBranch, string title, string description, string token)
    {
        var request = new HttpRequestMessage(HttpMethod.Post,
            $"https://gitlab.com/api/v4/projects/{projectPath}/merge_requests");
        
        request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", token);
        request.Content = new StringContent(
            JsonSerializer.Serialize(new
            {
                source_branch = sourceBranch,
                target_branch = targetBranch,
                title = title,
                description = description
            }),
            Encoding.UTF8,
            "application/json");

        var response = await _httpClient.SendAsync(request);
        response.EnsureSuccessStatusCode();

        var json = JObject.Parse(await response.Content.ReadAsStringAsync());
        return json["iid"]?.Value<int>() ?? 0;
    }

    /// <summary>
    /// Проверяет валидность токена
    /// </summary>
    public async Task<bool> ValidateTokenAsync(string token)
    {
        var request = new HttpRequestMessage(HttpMethod.Get, "https://gitlab.com/api/v4/user");
        request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", token);

        var response = await _httpClient.SendAsync(request);
        return response.IsSuccessStatusCode;
    }

    /// <summary>
    /// Получает информацию о пользователе по токену
    /// </summary>
    public async Task<string> GetUserNameAsync(string token)
    {
        var request = new HttpRequestMessage(HttpMethod.Get, "https://gitlab.com/api/v4/user");
        request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", token);

        var response = await _httpClient.SendAsync(request);
        response.EnsureSuccessStatusCode();

        var json = JObject.Parse(await response.Content.ReadAsStringAsync());
        return json["username"]?.ToString() ?? "unknown";
    }

    /// <summary>
    /// Получает название ветки по умолчанию для репозитория
    /// </summary>
    public async Task<string> GetDefaultBranchAsync(string owner, string repo, string? token = null)
    {
        var projectPath = EncodeProjectPath(owner, repo);
        var request = new HttpRequestMessage(HttpMethod.Get, 
            $"https://gitlab.com/api/v4/projects/{projectPath}");
        
        if (!string.IsNullOrEmpty(token))
            request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", token);

        var response = await _httpClient.SendAsync(request);
        response.EnsureSuccessStatusCode();

        var json = JObject.Parse(await response.Content.ReadAsStringAsync());
        return json["default_branch"]?.ToString() ?? "main";
    }
}

File: CodeQualityCheckerPython/services/gitlab_service.py
"""
Сервис для работы с GitLab API.
"""
import base64
from datetime import datetime
from typing import List, Optional, Tuple
from urllib.parse import quote, urlparse

import requests

from models import GitCommitResult, GitFileInfo, TokenValidationResult


class GitLabService:
    def __init__(self):
        self.session = requests.Session()

    def parse_gitlab_url(self, url: str) -> Tuple[str, str, str, str]:
        url = url.strip().rstrip("/")
        if url.endswith(".git"):
            url = url[:-4]
        parsed = urlparse(url)
        segments = [s for s in parsed.path.split("/") if s]
        if len(segments) < 2:
            raise ValueError("Неверный формат URL GitLab репозитория")
        owner, repo = segments[0], segments[1]
        branch, path = "", ""
        full_path = "/".join(segments)
        if "-/tree/" in full_path:
            after = full_path.split("-/tree/", maxsplit=1)[1]
            parts = after.split("/")
            branch = parts[0]
            path = "/".join(parts[1:]) if len(parts) > 1 else ""
        elif "-/blob/" in full_path:
            after = full_path.split("-/blob/", maxsplit=1)[1]
            parts = after.split("/")
            branch = parts[0]
            path = "/".join(parts[1:]) if len(parts) > 1 else ""
        return owner, repo, branch, path

    def _encode_project_path(self, owner: str, repo: str) -> str:
        return quote(f"{owner}/{repo}", safe="")

    async def get_default_branch(
        self, owner: str, repo: str, token: Optional[str] = None
    ) -> str:
        project_path = self._encode_project_path(owner, repo)
        headers = {"Authorization": f"Bearer {token}"} if token else {}
        response = self.session.get(
            f"https://gitlab.com/api/v4/projects/{project_path}", headers=headers
        )
        response.raise_for_status()
        return response.json().get("default_branch", "main")

    async def get_file_content(
        self, owner: str, repo: str, path: str, branch: str, token: Optional[str] = None
    ) -> str:
        project_path = self._encode_project_path(owner, repo)
        encoded_path = quote(path, safe="")
        headers = {"Authorization": f"Bearer {token}"} if token else {}
        url = (
            f"https://gitlab.com/api/v4/projects/{project_path}/repository/files/"
            f"{encoded_path}?ref={branch}"
        )
        response = self.session.get(url, headers=headers)
        response.raise_for_status()
        content = response.json().get("content", "")
        if not content:
            raise ValueError("Не удалось получить содержимое файла")
        return base64.b64decode(content).decode("utf-8", errors="replace")

    async def get_directory_contents(
        self, owner: str, repo: str, path: str, branch: str, token: Optional[str] = None
    ) -> List[GitFileInfo]:
        project_path = self._encode_project_path(owner, repo)
        headers = {"Authorization": f"Bearer {token}"} if token else {}
        if path:
            url = (
                f"https://gitlab.com/api/v4/projects/{project_path}/repository/tree"
                f"?path={quote(path, safe='')}&ref={branch}"
            )
        else:
            url = (
                f"https://gitlab.com/api/v4/projects/{project_path}/repository/tree"
                f"?ref={branch}"
            )
        response = self.session.get(url, headers=headers)
        response.raise_for_status()
        return [
            GitFileInfo(
                name=item.get("name", ""),
                path=item.get("path", ""),
                type="dir" if item.get("type") == "tree" else "file",
                size=0,
            )
            for item in response.json()
        ]

    async def create_commit_with_fixes(
        self,
        owner: str,
        repo: str,
        branch: str,
        file_path: str,
        fixed_content: str,
        commit_message: str,
        mr_description: str,
        token: str,
    ) -> GitCommitResult:
        if not token:
            return GitCommitResult(success=False, error_message="Требуется токен")
        project_path = self._encode_project_path(owner, repo)
        feature_branch = f"code-quality-fix-{datetime.now().strftime('%Y%m%d%H%M%S')}"
        headers = {"Authorization": f"Bearer {token}"}
        try:
            response = self.session.post(
                f"https://gitlab.com/api/v4/projects/{project_path}/repository/branches",
                headers=headers,
                json={"branch": feature_branch, "ref": branch},
            )
            response.raise_for_status()
            response = self.session.post(
                f"https://gitlab.com/api/v4/projects/{project_path}/repository/commits",
                headers=headers,
                json={
                    "branch": feature_branch,
                    "commit_message": commit_message,
                    "actions": [
                        {
                            "action": "update",
                            "file_path": file_path,
                            "content": fixed_content,
                        }
                    ],
                },
            )
            response.raise_for_status()
            response = self.session.post(
                f"https://gitlab.com/api/v4/projects/{project_path}/merge_requests",
                headers=headers,
                json={
                    "source_branch": feature_branch,
                    "target_branch": branch,
                    "title": commit_message.split("\n")[0],
                    "description": mr_description,
                },
            )
            response.raise_for_status()
            mr_data = response.json()
            return GitCommitResult(
                success=True,
                branch_name=feature_branch,
                pull_request_url=mr_data.get("web_url", ""),
                pull_request_number=mr_data.get("iid", 0),
            )
        except Exception as exc:
            return GitCommitResult(success=False, error_message=str(exc))

    async def add_merge_request_comment(
        self, owner: str, repo: str, mr_iid: int, comment: str, token: str
    ) -> bool:
        if not token:
            return False
        project_path = self._encode_project_path(owner, repo)
        response = self.session.post(
            f"https://gitlab.com/api/v4/projects/{project_path}/merge_requests/{mr_iid}/notes",
            headers={"Authorization": f"Bearer {token}"},
            json={"body": comment},
        )
        return response.status_code == 201

    async def validate_token(self, token: str) -> TokenValidationResult:
        try:
            response = self.session.get(
                "https://gitlab.com/api/v4/user",
                headers={"Authorization": f"Bearer {token}"},
            )
            if response.status_code == 200:
                return TokenValidationResult(
                    is_valid=True, user_name=response.json().get("username", "")
                )
            return TokenValidationResult(is_valid=False, error_message="Неверный токен")
        except Exception as exc:
            return TokenValidationResult(is_valid=False, error_message=str(exc))
